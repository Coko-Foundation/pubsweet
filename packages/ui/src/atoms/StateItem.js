import PropTypes from 'prop-types'
import styled from 'styled-components'
import React from 'react'

const darkGrey = '#404040'
const lightGrey = '#b3b3b3'

const disabled = `
  color: ${lightGrey};
  cursor: default;

  &:hover {
    color: ${lightGrey};
  }
`

const Root = styled.span`
  cursor: pointer;
  font-family: var(--font-interface);
  font-size: 16px;
  font-style: italic;
  padding: 0;

  &:focus {
    outline: none;
  }

  &:hover {
    color: ${darkGrey};
    transition: 0.25s ease-in-out 0s;
  }

  ${props => (props.disabled ? disabled : '')};
`

const StateItem = ({ update, disabled, values, index }) => {
  const callUpdateWithNextIndex = () => {
    if (disabled) return
    const nextIndex = arrayShift(values, index)
    update(values[index], nextIndex)
  }

  const arrayShift = (array, i) => (i === array.length - 1 ? 0 : i + 1)

  return (
    <Root
      disabled={disabled}
      onClick={callUpdateWithNextIndex}
      onKeyPress={callUpdateWithNextIndex}
      role="button"
      tabIndex="0"
    >
      {values[index]}
    </Root>
  )
}

StateItem.propTypes = {
  disabled: PropTypes.bool,
  index: PropTypes.number.isRequired,
  update: PropTypes.func.isRequired,
  values: PropTypes.arrayOf(PropTypes.string).isRequired,
}

StateItem.defaultProps = {
  disabled: false,
}

export default StateItem
