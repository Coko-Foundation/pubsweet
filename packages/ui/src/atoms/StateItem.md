An interactive element which renders a text element from a given index of a provided array and upon click runs a provided update function, passing it the current text element and the index of the next item in the array, or 0 if the current item is the last.

If the update function is used to update the index, the component will cycle through the array on click.

```js
const initialState = {
  values: ['To Clean', 'Cleaning', 'Cleaned'],
  index: 0,
}

const update = (currentValue, nextIndex) => {
  setState({ index: nextIndex })
}
;<StateItem values={state.values} index={state.index} update={update} />
```

If the component is passed the disabled prop, the update function is not run:

```js
const initialState = {
  values: ['To Clean', 'Cleaning', 'Cleaned'],
  index: 0,
}

const update = (currentValue, nextIndex) => {
  setState({ index: nextIndex })
}
;<StateItem
  values={state.values}
  index={state.index}
  disabled={true}
  update={update}
/>
```
