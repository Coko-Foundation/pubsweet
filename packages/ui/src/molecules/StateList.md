A list of State Items separated by a delimiter.

```js
const current = {
  style: 0,
  edit: 0,
  clean: 0,
  review: 0,
}
const stateValues = {
  clean: ['To Clean', 'Cleaning', 'Cleaned'],
  edit: ['To Edit', 'Editing', 'Edited'],
  review: ['To Review', 'Reviewing', 'Reviewed'],
  style: ['To Style', 'Styling', 'Styled'],
}
const update = data => {
  console.log('data', data)
}
;<StateList currentValues={current} values={stateValues} update={update} />
```
