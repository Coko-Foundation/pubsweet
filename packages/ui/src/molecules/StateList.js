import PropTypes from 'prop-types'
import React from 'react'
import { map, uniqueId, keys, last } from 'lodash'
import { ChevronRight } from 'react-feather'

import classes from './StateList.local.scss'
import StateItem from '../atoms/StateItem'

const StateList = ({ currentValues, update, values }) => {
  const progressIds = keys(values)
  const lastItem = last(progressIds)

  // TODO: Placeholder -- to be implemented with authsome
  const canAct = key => true

  const handleUpdate = (currentItem, nextIndex) => {
    update(currentItem, nextIndex)
  }

  const items = map(values, (valueList, currentItem) => {
    let delimiter
    const currentValueIndex = currentValues[currentItem]

    if (currentItem !== lastItem) {
      delimiter = <ChevronRight className={classes.delimiter} size={16} />
    }

    return (
      <div className={classes.itemContainer} key={uniqueId()}>
        <StateItem
          disabled={!canAct(currentItem)}
          index={currentValueIndex}
          update={handleUpdate}
          values={valueList}
        />
        {delimiter}
      </div>
    )
  })

  return <div className={classes.stateListContainer}>{items}</div>
}

StateList.propTypes = {
  currentValues: PropTypes.objectOf(PropTypes.number).isRequired,
  update: PropTypes.func.isRequired,
  values: PropTypes.objectOf(PropTypes.arrayOf(PropTypes.string)).isRequired,
}

export default StateList
