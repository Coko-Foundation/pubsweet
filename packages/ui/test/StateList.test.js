import React from 'react'
import { forIn } from 'lodash'
import { shallow } from 'enzyme'
import 'jest-styled-components'
import renderer from 'react-test-renderer'

import StateItem from '../src/atoms/StateItem'
import StateList from '../src/molecules/StateList'

const currentValues = {
  style: 0,
  edit: 0,
  clean: 0,
  review: 0,
}
const stateValues = {
  clean: ['To Clean', 'Cleaning', 'Cleaned'],
  edit: ['To Edit', 'Editing', 'Edited'],
  review: ['To Review', 'Reviewing', 'Reviewed'],
  style: ['To Style', 'Styling', 'Styled'],
}

const props = {
  currentValues,
  values: stateValues,
  update: () => null,
}
const wrapper = shallow(<StateList {...props} />)
const stateItems = wrapper.find(StateItem)

describe('StateList', () => {
  test('is rendered correctly', () => {
    const tree = renderer.create(<StateList {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('should contain four State Item children', () => {
    const itemsNumber = Object.keys(stateValues).length

    expect(stateItems.exists()).toEqual(true)
    expect(stateItems).toHaveLength(itemsNumber)
  })

  test('gets the correct props', () => {
    let i = 0
    const stateItemComp = stateItems.getElements()

    forIn(stateValues, (value, key) => {
      const stateItem = stateItemComp[i]
      const stateItemProps = stateItem.props

      expect(stateItemProps.disabled).toEqual(false)
      expect(stateItemProps.index).toEqual(props.currentValues[key])
      expect(stateItemProps.values).toEqual(props.values[key])

      i += 1
    })
  })
})
