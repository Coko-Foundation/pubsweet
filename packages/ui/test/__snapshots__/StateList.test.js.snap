// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateList is rendered correctly 1`] = `
.c0 {
  cursor: pointer;
  font-family: var(--font-interface);
  font-size: 16px;
  font-style: italic;
  padding: 0;
}

.c0:focus {
  outline: none;
}

.c0:hover {
  color: #404040;
  -webkit-transition: 0.25s ease-in-out 0s;
  transition: 0.25s ease-in-out 0s;
}

<div
  className="stateListContainer"
>
  <div
    className="itemContainer"
  >
    <span
      className="c0"
      disabled={false}
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      To Clean
    </span>
    <svg
      className="delimiter"
      fill="none"
      height={16}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="9 18 15 12 9 6"
      />
    </svg>
  </div>
  <div
    className="itemContainer"
  >
    <span
      className="c0"
      disabled={false}
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      To Edit
    </span>
    <svg
      className="delimiter"
      fill="none"
      height={16}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="9 18 15 12 9 6"
      />
    </svg>
  </div>
  <div
    className="itemContainer"
  >
    <span
      className="c0"
      disabled={false}
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      To Review
    </span>
    <svg
      className="delimiter"
      fill="none"
      height={16}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={16}
      xmlns="http://www.w3.org/2000/svg"
    >
      <polyline
        points="9 18 15 12 9 6"
      />
    </svg>
  </div>
  <div
    className="itemContainer"
  >
    <span
      className="c0"
      disabled={false}
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      tabIndex="0"
    >
      To Style
    </span>
  </div>
</div>
`;
